File: 215-kth-largest-element-in-an-array/215-kth-largest-element-in-an-array.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        k-=1
        return self.__finder(nums, 0,len(nums)-1, k)
    def __finder(self,nums:list, start:int, end:int, k:int)->int:
        if start==end:
            return nums[start]
        while start<end:
            q = self.__partition(nums, start, end)
            if k>q: # k-th element is bigger than partition index
                return self.__finder(nums, q+1,end, k)

            elif q>k:
                return self.__finder(nums, start, q-1, k)
            else:
                return nums[q]

    def __partition(self,l: list, start: int, end: int) -> int:
        j = start
        for i in range(start + 1, end + 1):
            if l[i] > l[start]:
                j += 1
                l[j], l[i] = l[i], l[j]
        l[start], l[j] = l[j], l[start]
        return j

File: 215-kth-largest-element-in-an-array/215-kth-largest-element-in-an-array.cpp
#include <iostream>
#include <vector>
using namespace std;
using std::swap;
int findKthelement(vector<int>& nums, int k);
int finder(vector<int>& nums, int start, int end, int k);
int partition(vector<int>& nums, int start, int end);

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        k--;
        int len = nums.size();
        return finder(nums, 0, len-1, k);
    }
    int finder(vector<int>& nums, int start, int end, int k) {
        if (start == end) return nums[start];
        while (start < end) {
            int q = partition(nums, start, end);
            if (k > q) return finder(nums, q + 1, end, k);
            else if (q > k) return finder(nums, start, q - 1, k);
            else return nums[q];
        }
        return 0;
    }
    int partition(vector<int>& nums, int start, int end) {
        int j = start;
        for (int i = start + 1; i < (end + 1); i++) {
            if (nums[i] > nums[start]) {
                j++; //omit later
                swap(nums[j], nums[i]);
            }
        }
    swap(nums[start], nums[j]);
    return j;
    }

  
};